### PR TITLE
pml_ucx: add ompi datatype attribute to release ucp_datatype

### PR DESCRIPTION
**Pull request by yosefe**
Fixes openucx/ucx#2921
(checked that the leak is gone with IMB Allgatherv 512b, and passed mpi_test_suite)

### CODE REVIEW

**Comment by yosefe** [comment]
adding @ca-taylor

**Comment by yosefe** [comment]
@vspetrov @hoopoepg @brminich can you pls take a look?

File: ompi/mca/pml/ucx/pml_ucx.c [hunk 1]
```diff
@@ -249,14 +254,33 @@ int mca_pml_ucx_init(void)
                     (void *)ompi_pml_ucx.ucp_context,
                     (void *)ompi_pml_ucx.ucp_worker);
     return OMPI_SUCCESS;
+
+err_destroy_worker:
+    ucp_worker_destroy(ompi_pml_ucx.ucp_worker);
+    ompi_pml_ucx.ucp_worker = NULL;
+err:
+    return OMPI_ERROR;
```
Comment (brminich) [inline ompi/mca/pml/ucx/pml_ucx.c]: `return rc`

**Comment by AdamSimpson** [comment]
Will this make it into the 4.0 release?

File: ompi/mca/pml/ucx/pml_ucx_datatype.c [hunk 2]
```diff
@@ -147,16 +161,33 @@ ucp_datatype_t mca_pml_ucx_init_datatype(ompi_datatype_t *datatype)
...
+    datatype->pml_data = ucp_datatype;
+
+    /* Add custom attribute, to clean up UCX resources when OMPI datatype is
+     * released.
+     */
+    if (ompi_datatype_is_predefined(datatype)) {
+        PML_UCX_ASSERT(datatype->id < OMPI_DATATYPE_MAX_PREDEFINED);
+        ompi_pml_ucx.predefined_types[datatype->id] = ucp_datatype;
+    } else {
+        ret = ompi_attr_set_c(TYPE_ATTR, datatype, &datatype->d_keyhash,
```
[the remark this answers is not in the document]
Comment (yosefe) [reply ompi/mca/pml/ucx/pml_ucx_datatype.c]: would the delete callback be called for an mpi datatype if i don't set my attribute? 
according to the code of omi_attr_delete_impl - no:
https://github.com/open-mpi/ompi/blob/master/ompi/attribute/attribute.c#L1055

**Comment by yosefe** [comment]
@AdamSimpson yes the plan is to backport to v4.0.x as well

File: ompi/mca/pml/ucx/pml_ucx_datatype.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (vspetrov) [reply ompi/mca/pml/ucx/pml_ucx_datatype.c]: right, i see

**Comment by akesandgren** [comment]
I see no trace of this fix in the 3.1.x or the 3.0.x branches, would probably be good to have it there too.

**Comment by jsquyres** [comment]
@yosefe @brminich @vspetrov We are aiming for the end of the month for v3.0.x and v3.1.x releases.  If you have small bug fixes that need to get to these release branches, you should do so ASAP.

**Comment by brminich** [comment]
@akesandgren, is it critical for you? Can you use OMPI 4.0x?

**Comment by akesandgren** [comment]
We use everything from 1.6 to 3.1.4 at the moment, 4.x is in the pipeline (but at least ScaLAPACK need to get some fixes).
We in this case is the community that uses EasyBuild to install HPC software.
We already have this PR as a patch to our builds but it would be better to not have to do that.

So no, it's not critical, just highly preferable.

**Pull request merged**